recorder/src/main/cpp/controller.cc

```diff
@@ -290,56 +282,260 @@ void Controller::run() {
     // close(clientConnection);
 }
 
-void Controller::with_current_work(std::function<void(Controller::W&, Controller::WSt&, Controller::WRes&, std::string&)> proc) {
-    std::lock_guard<std::mutex> current_work_guard(current_work_mtx);
-    proc(current_work, current_work_state, current_work_result, current_work_description);
+void Controller::with_current_work(std::function<void(Controller::W&, Controller::WSt&, Controller::WRes&, Time::Pt&, Time::Pt&)> proc) {
+    std::lock_guard<std::recursive_mutex> current_work_guard(current_work_mtx);
+    proc(current_work, current_work_state, current_work_result, work_start, work_end);
 }
 
-void Controller::accept_work(Buff& poll_response_buff) {
+void Controller::accept_work(Buff& poll_response_buff, const std::string& host, const std::uint32_t port) {
     recording::PollRes res;
+    logger->trace("Accept-work host: {} and port: {}", host, port);
     auto parse_successful = res.ParseFromArray(poll_response_buff.buff + poll_response_buff.read_end, poll_response_buff.write_end - poll_response_buff.read_end);
     if (! parse_successful) {
         logger->error("Parse of poll-response failed, ignoring it");
         return;
     }
-    
-    with_current_work([&res](Controller::W& w, Controller::WSt& wst, Controller::WRes& wres, std::string& desc) {
-            auto new_work = res.mutable_assignment();
-            if (wst != recording::WorkResponse::complete) {
-                logger->critical("New work (id: {}, desc: {}, controller: {}/{}) issued while current-work (id: {}, desc: {}) is incomplete (state {})", 
-                                 new_work->work_id(), res.work_description(), res.controller_id(), res.controller_version(),
-                                 w.work_id(), wst, desc);
-                return;
-            }
+    if (res.has_assignment()) {
+        with_current_work([&](Controller::W& w, Controller::WSt& wst, Controller::WRes& wres, Time::Pt& start_tm, Time::Pt& end_tm) {
+                auto new_work = res.mutable_assignment();
+
+                if (wst != recording::WorkResponse::complete) {
+                    logger->critical("New work (id: {}, desc: {}, controller: {}/{}) issued while current-work (id: {}, desc: {}) is incomplete (state {})", 
+                                     new_work->work_id(), new_work->description(), res.controller_id(), res.controller_version(),
+                                     w.work_id(), wst, w.description());
+                    return;
+                }
+
+                logger->trace("New work (id: {}, desc: {}, controller: {}/{}) is being assigned",
+                              new_work->work_id(), new_work->description(), res.controller_id(), res.controller_version());
+
+                w.Swap(new_work);
+
+                const auto delay = w.delay();
+
+                start_tm = end_tm = Time::now();
+
+                if ((delay + w.duration()) > 0) {
```

Comment:
hmm, so here 0 sec duration means execute/doWork once. I am assuming that work spans over a period.